### PR TITLE
Add support of state hash calculation into runtimes & services

### PR DESCRIPTION
**Pull request by popzxc**
### Definition of Done

[coding guidelines]: https://github.com/exonum/exonum/blob/master/CONTRIBUTING.md#conventions

### CODE REVIEW

File: exonum/src/blockchain/mod.rs [hunk 1]
```diff
@@ -349,16 +349,16 @@ impl Blockchain {
...
-                    //     let service_id = service.service_id();
-                    //     let vec_service_state = service.state_hash(&fork);
-                    //     for (idx, service_table_hash) in vec_service_state.into_iter().enumerate() {
-                    //         let key = Self::service_table_unique_key(service_id, idx);
-                    //         state_hashes.push((key, service_table_hash));
-                    //     }
-                    // }
+                    let dispatcher = self.dispatcher.lock().expect("Expected lock on Dispatcher");
+
+                    for hashes in dispatcher.state_hashes(&fork) {
```
Comment (aleksuss) [inline exonum/src/blockchain/mod.rs]:
```suggestion
                    for (service_id, vec_service_state) in dispatcher.state_hashes(&fork) {
```

Comment (aleksuss) [reply exonum/src/blockchain/mod.rs]: And 355 and 356 line could be removed ?

Comment (popzxc) [reply exonum/src/blockchain/mod.rs]: Done.

File: exonum/src/runtime/dispatcher.rs [hunk 2]
```diff
@@ -142,6 +143,16 @@ impl RuntimeEnvironment for Dispatcher {
...
 
+    fn state_hashes(&self, snapshot: &dyn Snapshot) -> Vec<(ServiceInstanceId, Vec<Hash>)> {
+        let mut hashes = Vec::new();
+
+        for (_, runtime) in &self.runtimes {
+            hashes.append(&mut runtime.state_hashes(snapshot));
+        }
+
+        hashes
+    }
```
Comment (aleksuss) [inline exonum/src/runtime/dispatcher.rs]: Or you can re-write it for something like this:
```rust
fn state_hashes(&self, snapshot: &dyn Snapshot) -> Vec<(ServiceInstanceId, Vec<Hash>)> {
        self.runtimes
            .iter()
            .map(|(_, runtime)| runtime.state_hashes(snapshot))
            .flatten()
            .collect::<Vec<_>>()
}
```

Comment (popzxc) [reply exonum/src/runtime/dispatcher.rs]: Hmm, dunno. I like plain&dumb style here more. However, if you insist...

Comment (aleksuss) [reply exonum/src/runtime/dispatcher.rs]: Not really but IMO looks better.

**Pull request merged**